### PR TITLE
[DEM-582] FSP and non-FSP in projectq

### PR DESCRIPTION
**Pull request by QFer**
* When measurements are part of the circuit, the circuit is executed as non-FSP, otherwise the circuit is executed as FSP.
* Added check for invalid number of runs (must be > 0).

### CODE REVIEW

File: src/tests/quantuminspire/projectq/test_backend_qx.py [hunk 1]
```diff
@@ -160,6 +165,23 @@ def test_store_returns_correct_qasm(self):
         self.__store_function_assert_equal(0, T, "\nt q[0]")
         self.__store_function_assert_equal(0, Tdag, "\ntdag q[0]")
 
+    @patch('quantuminspire.projectq.backend_qx.get_control_count')
+    def __store_function(self, backend, identity, gate, function_mock, count=0, verbose=0):
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/tests/quantuminspire/projectq/test_backend_qx.py]: deleted it

File: src/tests/quantuminspire/projectq/test_backend_qx.py [hunk 2]
```diff
@@ -160,6 +165,23 @@ def test_store_returns_correct_qasm(self):
...
+        function_mock.return_value = count
+        command = MagicMock(gate=gate, qubits=[[MagicMock(id=identity)]],
+                            control_qubits=[MagicMock(id=identity - 1), MagicMock(id=identity)])
+        backend._store(command)
+
+    def test_store_returns_correct_qasm_program(self):
+        api = MockApiClient()
+        backend = QIBackend(quantum_inspire_api=api)
+        backend.main_engine = MagicMock(mapper=None)
+        angle = 0.1
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/tests/quantuminspire/projectq/test_backend_qx.py]: deleted it

File: src/tests/quantuminspire/projectq/test_backend_qx.py [hunk 3]
```diff
@@ -180,6 +202,34 @@ def test_store_deallocate_verbose_output(self):
...
+    def test_store_measure_gate_fsp(self, function_mock):
+        mock_tag = '0'
+        api = MockApiClient()
+        function_mock.return_value = 4
+        backend = QIBackend(quantum_inspire_api=api)
+        command = MagicMock(gate=Measure, qubits=[[MagicMock(id=mock_tag)]],
+                            control_qubits=[MagicMock(id=2), MagicMock(id=3)],
+                            tags=[])
+        backend.main_engine = MagicMock(mapper=None)
+        backend._clear = False
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/tests/quantuminspire/projectq/test_backend_qx.py]: I replaced them with public logic

**Comment by QFer** [comment]
@qSaevar you're right about this circuit being non-fsp but I think it's a corner case where linear mapper is demonstrated with few supported gates, generating a lot of alternative gates. When CNOT is added to two_qubit_gates (and optional H as one_qubit_gate the result will be fsp. Also when running without the mapper this circuit would be executed as fsp.
